### PR TITLE
feat(api): API for updating File copyright content

### LINKED ISSUE #2468: Copyright File-view
File Details (Copyright/email/url/Author, ECC, IPRA) Content, Count, finding, activated/deactivated, delete + User findings with  description, text finding, comment for both activated and deactivated. With all decision options

### PR DESCRIPTION
**Pull request by soham4abc**
## Description

API to update File copyright is added  in this PR

## How to test

Send a `PUT` request to the endpoint `/file/upload/{Upload_id}/item/{Item_id}/copyright/{Copyright_hash}`

with a body :
```
{
  "content": "New Test copyright text"
}
```
##Screenshots:
- Before sending the API request:

Screenshot from 2023-06-18 18-09-36

- The API Request:

Screenshot from 2023-06-18 18-09-50

- After the API request:

Screenshot from 2023-06-18 18-10-01

closes #2468

Please consider using the closing keyword if the pull request is proposed to
fix an issue already created in the repository
(https://help.github.com/articles/closing-issues-using-keywords/)

<a href="https://gitpod.io/#https://github.com/fossology/fossology/pull/2479"></a>

### CODE REVIEW

**Comment by GMishx** [comment]
Would recommend to put the actual body content in the screenshots rather than just showing the response :-)

File: src/www/ui/api/Controllers/CopyrightController.php [hunk 1]
```diff
@@ -0,0 +1,112 @@
+<?php
+/*
+ SPDX-FileCopyrightText: © 2018 Siemens AG
+ Author: Gaurav Mishra <mishra.gaurav@siemens.com>
```
Comment (GMishx) [inline src/www/ui/api/Controllers/CopyrightController.php:4]: I did not write this file :smiley:

File: src/www/ui/api/Controllers/CopyrightController.php [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+  public function __construct($container)
+  {
+    parent::__construct($container);
+    $this->container = $container;
+    $this->clearingDao = $this->container->get('dao.clearing');
+    $this->licenseDao = $this->container->get('dao.license');
+    $this->uploadDao = $container->get('dao.upload');
+    $this->restHelper = $container->get('helper.restHelper');
+    $this->copyrightDao = $container->get('dao.copyright');
+    $this->CopyrightHist = $this->restHelper->getPlugin('ajax-copyright-hist');
```
Comment (GMishx) [inline src/www/ui/api/Controllers/CopyrightController.php:72]: Check review comments from https://github.com/fossology/fossology/pull/2478#discussion_r1246464309

File: src/www/ui/api/Controllers/CopyrightController.php [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+   *
+   * @param  ServerRequestInterface $request
+   * @param  ResponseHelper         $response
+   * @param  array                  $args
+   * @return ResponseHelper
+   */
+
+  public function UpdateFileCopyrights($request, $response, $args)
+  {
+    try {
```
Comment (GMishx) [inline src/www/ui/api/Controllers/CopyrightController.php:86]: Limit the scope of `try/catch` block.

File: src/www/ui/api/Controllers/CopyrightController.php [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+  /**
+   * Update copyrights for a particular file
+   *
+   * @param  ServerRequestInterface $request
+   * @param  ResponseHelper         $response
+   * @param  array                  $args
+   * @return ResponseHelper
+   */
+
+  public function UpdateFileCopyrights($request, $response, $args)
```
Comment (GMishx) [inline src/www/ui/api/Controllers/CopyrightController.php:84]: Please following coding guidelines.
```suggestion
  public function updateFileCopyrights($request, $response, $args)
```

File: src/www/ui/api/Controllers/CopyrightController.php [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+  {
+    try {
+      $uploadDao = $this->restHelper->getUploadDao();
+      $uploadTreeId = intval($args['itemId']);
+      $copyrightHash = ($args['hash']);
```
Comment (GMishx) [inline src/www/ui/api/Controllers/CopyrightController.php:89]:
```suggestion
      $copyrightHash = $args['hash'];
```

File: src/www/ui/api/Controllers/CopyrightController.php [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+      $userId = $this->restHelper->getUserId();
+      $UploadTreeTableName = $uploadDao->getUploadtreeTableName($uploadTreeId);
```
Comment (GMishx) [inline src/www/ui/api/Controllers/CopyrightController.php:91]: Please following coding guidelines.
```suggestion
      $uploadTreeTableName = $uploadDao->getUploadtreeTableName($uploadTreeId);
```

File: src/www/ui/api/Controllers/CopyrightController.php [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+      $item = $uploadDao->getItemTreeBounds($uploadTreeId, $UploadTreeTableName);
+      $body = $this->getParsedBody($request);
+      $content = $body['content'];
+
+      if (!$this->dbHelper->doesIdExist($uploadDao->getUploadtreeTableName($uploadTreeId), "uploadtree_pk", $uploadTreeId)) {
+        $returnVal = new Info(404, "Item does not exist", InfoType::ERROR);
+      }
+      if ($returnVal !== null) {
+        return $response->withJson($returnVal->getArray(), $returnVal->getCode());
+      }
```
Comment (GMishx) [inline src/www/ui/api/Controllers/CopyrightController.php:103]: Check if the upload ID, item ID and copyright exists before performing operations.

File: src/www/ui/api/documentation/openapi.yaml [hunk 8]
```diff
@@ -1544,6 +1544,61 @@ paths:
                 $ref: '#/components/schemas/Info'
         default:
           $ref: '#/components/responses/defaultResponse'
+  /file/upload{id}/item/{itemId}/copyright/{hash}:
```
Comment (GMishx) [inline src/www/ui/api/documentation/openapi.yaml:1547]: Be consistent with other endpoints
```suggestion
  /uploads/{id}/item/{itemId}/copyrights/{hash}:
```

File: src/www/ui/api/documentation/openapi.yaml [hunk 9]
```diff
@@ -2640,6 +2695,27 @@ components:
...
+          description: Status code
+          example: 200
+        content:
+          type: string
+          description: Success message
+          example: "Successfully Updated Copyright."
+        type:
+          type: string
+          description: type of message
+          example: "INFO"
```
Comment (GMishx) [inline src/www/ui/api/documentation/openapi.yaml:2712]: The response code is exactly same as `INFO`?

**Review by GMishx — CHANGES_REQUESTED** [review]

**Comment by soham4abc** [comment]
> Would recommend to put the actual body content in the screenshots rather than just showing the response :-)

Sure sir, I will be updating the PR accordingly soon.

File: src/www/ui/api/Controllers/CopyrightController.php [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+    $copyrightHash = $args['hash'];
+    $userId = $this->restHelper->getUserId();
+    $cpTable = $this->copyrightHist->getTableName('statement');
+    $returnVal = null;
+    $body = $this->getParsedBody($request);
+    $content = $body['content'];
+    try {
+      if (!$this->dbHelper->doesIdExist($this->restHelper->getUploadDao()->getuploadTreeTableName($uploadTreeId), "uploadtree_pk", $uploadTreeId)) {
+        $returnVal = new Info(404, "Item does not exist", InfoType::ERROR);
+      }
```
Comment (GMishx) [inline src/www/ui/api/Controllers/CopyrightController.php:75]: Please add a sanity check for upload ID as well, even though the backend doesn't need it.

**Review by GMishx — CHANGES_REQUESTED** [review]

File: src/www/ui/api/Controllers/CopyrightController.php (hunk 10 above)
Comment (soham4abc) [reply src/www/ui/api/Controllers/CopyrightController.php:75]: Sure I will add it

File: src/www/ui/api/Controllers/CopyrightController.php [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+  /**
+   * @var CopyrightDao $copyrightDao
+   * Copyright Dao object
+   */
+  private $copyrightDao;
+
+  public function __construct($container)
+  {
+    parent::__construct($container);
+    $this->restHelper = $container->get('helper.restHelper');
```
Comment (GMishx) [inline src/www/ui/api/Controllers/CopyrightController.php:48]: Please remove the assignment. It is already done for you in the parent class.

File: src/www/ui/api/Controllers/CopyrightController.php [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+    $this->copyrightDao = $container->get('dao.copyright');
```
Comment (GMishx) [inline src/www/ui/api/Controllers/CopyrightController.php:49]:
```suggestion
    $this->copyrightDao = $this->container->get('dao.copyright');
```

File: src/www/ui/api/documentation/openapi.yaml [hunk 13]
```diff
@@ -1615,6 +1615,61 @@ paths:
...
+            schema:
+              $ref: '#/components/schemas/SetCopyrightInfo'
+      responses:
+        '200':
+          description: OK
+          headers:
+            X-Total-Pages:
+              description: Total number of pages which can be generated based on limit
+              schema:
+                type: integer
```
Comment (GMishx) [inline src/www/ui/api/documentation/openapi.yaml:1658]: The endpoint is not generating this header in the code and is not needed as well.

**Review by GMishx — CHANGES_REQUESTED** [review]

**Comment by soham4abc** [comment]
Made the requested changes sir,

Please do have a look

@GMishx

**Review by GMishx — CHANGES_REQUESTED** [review]

**Comment by soham4abc** [comment]
Made the requested changes sir.
Please do have a look @GMishx

**Review by GMishx — APPROVED** [review]
Changes looks good.

**Review by GMishx — APPROVED** [review]
Tested, working as expected.

**Comment by GMishx** [comment]
@soham4abc , kindly rebase for merge.

**Comment by soham4abc** [comment]
I'm rebasing the branch

**Comment by soham4abc** [comment]
@GMishx Please do have a look

**Pull request merged**